### PR TITLE
[Merged by Bors] - feat: port Logic.Function.Conjugate

### PR DESCRIPTION
**Pull request by joneugster**
This is my first PR to mathlib4. It seems to be a short and simple file, but could you please check I got the naming conventions right, please? (in particular: is it `Semiconj` or `SemiConj`?)

No name changes from mathlib3port, only had to adapt some proofs:
e.g. `function.comp_app` is now `Function.comp_apply`. Does this need an `#align` statement, too?

### CODE REVIEW

**Comment by kbuzzard** [comment]
This looks good to me (modulo trivial simp only duplication comment). Remark: I reviewed this by porting the file myself and then comparing; my conclusion is that Jon is better at spotting dostrings which need to be updated than I do.

File: Mathlib/Logic/Function/Conjugate.lean [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+protected theorem comp_eq (h : Semiconj f ga gb) : f ∘ ga = gb ∘ f :=
+  funext h
+
+protected theorem eq (h : Semiconj f ga gb) (x : α) : f (ga x) = gb (f x) :=
+  h x
+
+theorem comp_right (h : Semiconj f ga gb) (h' : Semiconj f ga' gb') :
+    Semiconj f (ga ∘ ga') (gb ∘ gb') :=
+  fun x => by
+    simp only [comp_apply, comp_apply, h.eq, h'.eq]
```
Comment (kbuzzard) [inline Mathlib/Logic/Function/Conjugate.lean:46]:
```suggestion
simp only [comp_apply, h.eq, h'.eq]
```

**Review by kbuzzard — APPROVED** [review]

**Comment by kbuzzard** [comment]
Re: `Function.comp_apply`. I believe you're right, this needs an #align. Because the definition is in core Lean 3 I don't know where to put it -- perhaps there is some "core Lean 3 #align ing stuff" file somewhere which it should go in, which I might have seen in the past and then forgotten where it was?

**Comment by digama0** [comment]
> perhaps there is some "core Lean 3 #align ing stuff" file somewhere which it should go in, which I might have seen in the past and then forgotten where it was? 

That would be `Mathlib.Init.Align`, but it is only for files that are nothing but `#align`s / unported files. If the lean 3 companion file exists then just put the `#align` there.

File: Mathlib/Logic/Function/Conjugate.lean [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+theorem Semiconj.commute {f g : α → α} (h : Semiconj f g g) : Commute f g :=
+  h
+
+namespace Commute
+
+variable {f f' g g' : α → α}
+
+@[refl]
+theorem refl (f : α → α) : Commute f f :=
+  fun _x => Eq.refl _
```
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Logic/Function/Conjugate.lean:87]: _

**Comment by semorrison** [comment]
Since `function.comp_app` was defined in Lean3 in `init/function.lean`, the #align statement should go in `Mathlib/Init/Function.lean`. I've just added it now.

File: Mathlib/Logic/Function/Conjugate.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (semorrison) [reply Mathlib/Logic/Function/Conjugate.lean:87]: I've done this change.

**Comment by semorrison** [comment]
bors merge

**Comment by joneugster** [comment]
> Re: `Function.comp_apply`. I believe you're right, this needs an `#align function.comp_app Function.comp_apply` somewhere. Because the definition is in core Lean 3 I don't know where to put the alignment -- perhaps there is some "core Lean 3 #align ing stuff" file somewhere which it should go in, which I might have seen in the past and then forgotten where it was? My instinct is that it's better to get this alignment done ASAP (i.e. in this PR) because it makes the porting process easier for others (ultimately I am hoping for a time where even beginners will be able to contribute because the machine translation is so good).

Thanks @kbuzzard for the detailed comments!

**Pull request closed without merging**